[FLINK-16887][table-planner-blink] Refactor retraction rules to support inferring ChangelogMode

wuchong opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Current retraction machanism to support more message kinds (insert/delete/update_before/update_after). And make the new machanism more extensible for future features and more readable. 

## Brief change log

Design doc: Design documentation: https://docs.google.com/document/d/1n_iXIQsKT3uiBqENR8j8RdjRhZfzMhhB66QZvx2rFjE/edit?ts=5e8419c1#

- 1st commit: introduce `FlinkChangelogModeInferenceProgram` with 2 traits and 2 satisfy visitor to infer changelog mode for every nodes.
- 2nd commit: Remove interfaces and traits and rules of legacy retraction mechanism entirely
- 3rd commit: Use new changelog trait description in plan tests and update all plan xml files
- 4th commit: Fix remaining tests because of the validation change and plan change

## Verifying this change

This pull request is verified with existing tests. 

But some tests are modified to have the new changelog trait description instead of the legacy AccMode trait.

The plan change in `DagOptimizationTest` is reasonable, because we can reuse more nodes after refactoring. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

wuchong commented on flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/DagOptimizationTest.xml:
```diff
@@ -959,15 +951,15 @@ LogicalSink(name=[`default_catalog`.`default_database`.`upsertSink`], fields=[b,
-      +- GroupAggregate(groupBy=[c], select=[c, SUM(a) AS a_sum], updateAsRetraction=[false], accMode=[Acc])
-         +- Exchange(distribution=[hash[c]], updateAsRetraction=[true], accMode=[Acc])
-            +- Calc(select=[c, a], updateAsRetraction=[true], accMode=[Acc])
+Sink(name=[`default_catalog`.`default_database`.`upsertSink`], fields=[b, c, a_sum], changelogMode=[NONE])
```
Since FLIP-95 new sink interfaces, PK of sink is defined in DDL, not derived from query anymore. Deriving primary key from query confused many users and make Flink SQL not flexible. So I will add the PK information once we integrate with new sink interface.

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/trait/UpdateKindTrait.scala:
```diff
@@ -0,0 +1,79 @@
+}
+
+object UpdateKindTrait {
+  val NO_UPDATE = new UpdateKindTrait(UpdateKind.NO_UPDATE)
```
`NO_UPDATE` is not just a placeholder here. For `AppendTableSink` and window aggregation, it requires children should produce `NO_UPDATE` trait. It doesnt make sense to require a `ONLY_UPDATE_AFTER` or `BEFORE_AND_AFTER` in this case. 

Merging `ONLY_UPDATE_AFTER` and `BEFORE_AND_AFTER` into `ModifyKind.UPDATE` maybe confusing. Becasue a default `BEFORE_AND_AFTER` maybe wrong in some cases (e.g. upsert source produce `ONLY_AFTER`). Separating into 2 traits in on purpose, and I think it makes things cleaner.

wuchong replied:
I will add some comments here.

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala:
```diff
@@ -0,0 +1,595 @@
+        val rankStrategies = RankProcessStrategy.analyzeRankProcessStrategies(
+          rank.getInput, rank.partitionKey, rank.orderKey)
+        // creates new Rank nodes for every applied RankStrategy
+        val newRankNodes = rankStrategies.map(rank.copy)
```
We can't move it into `visitRankStrategies`, because `rank.copy(RankProcessStrategy)` is not a common interface of `StreamExecRank` and `StreamExecSortLimit`. I'm not sure whether we should introduce a common abstract for `StreamExecRank` and `StreamExecSortLimit`.

wuchong replied:
I added a `applyRankStrategy: RankProcessStrategy => StreamPhysicalRel` paramter to the method `visitRankStrategies` to make it possible move rank.copy into this method.

wuchong commented on flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/rules/physical/stream/RetractionRulesWithTwoStageAggTest.xml:
```diff
@@ -39,19 +39,19 @@ LogicalAggregate(group=[{0}], frequency=[COUNT($0)])
-               +- TableSourceScan(table=[[default_catalog, default_database, MyTable2, source: [TestTableSource(word, cnt)]]], fields=[word, cnt], updateAsRetraction=[true], accMode=[Acc])
+GlobalGroupAggregate(groupBy=[cnt], select=[cnt, COUNT_RETRACT(count$0) AS frequency], changelogMode=[I,UA,D])
++- Exchange(distribution=[hash[cnt]], changelogMode=[I])
+   +- LocalGroupAggregate(groupBy=[cnt], select=[cnt, COUNT_RETRACT(cnt) AS count$0, COUNT_RETRACT(*) AS count1$1], changelogMode=[I])
```
It is as expected. LocalAggregate will emit accumulators to global aggregates. All the accumulators are encoded in insert messages.

wuchong commented on flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/rules/physical/stream/RetractionRulesTest.xml:
```diff
@@ -39,15 +39,15 @@ LogicalAggregate(group=[{0}], frequency=[COUNT($0)])
-      :        +- TableSourceScan(table=[[default_catalog, default_database, MyTable, source: [TestTableSource(word, number)]]], fields=[word, number], updateAsRetraction=[true], accMode=[Acc])
-      +- Calc(select=[cnt], updateAsRetraction=[true], accMode=[Acc])
-         +- TableSourceScan(table=[[default_catalog, default_database, MyTable2, source: [TestTableSource(word, cnt)]]], fields=[word, cnt], updateAsRetraction=[true], accMode=[Acc])
+GroupAggregate(groupBy=[cnt], select=[cnt, COUNT_RETRACT(cnt) AS frequency], changelogMode=[I,UA,D])
```
I merged `RetractionRulesTest` and `RetractionRulesWithTwoStageAggTest` into `ChangelogModeInferenceTest` using `Parameterized`.

wuchong commented on flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/planner/plan/stream/sql/join/JoinTest.xml:
```diff
@@ -313,17 +313,17 @@ LogicalProject(a1=[$1], b1=[$3])
-      +- Exchange(distribution=[hash[b1]], updateAsRetraction=[true], accMode=[Acc])
-         +- Calc(select=[b1, b2], updateAsRetraction=[true], accMode=[Acc])
-            +- TableSourceScan(table=[[default_catalog, default_database, B, source: [TestTableSource(b1, b2, b3)]]], fields=[b1, b2, b3], updateAsRetraction=[true], accMode=[Acc])
+Join(joinType=[InnerJoin], where=[=(a1, b1)], select=[a1, b1], leftInputSpec=[JoinKeyContainsUniqueKey], rightInputSpec=[JoinKeyContainsUniqueKey], changelogMode=[I,UA,D])
```
Good catch!

wuchong commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/ChangelogPlanUtils.scala:
```diff
@@ -0,0 +1,116 @@
+   *  <p>Note: this method must be called after [[FlinkChangelogModeInferenceProgram]] is applied.
+   */
+  def isInsertOnly(plan: RelNode): Boolean = {
+    assert(plan.isInstanceOf[StreamPhysicalRel])
```
I still would like to keep this parameter as `RelNode`, otherwise, we will have a lot of duplicate codes, e.g. `getInput().asInstanceOf[StreamPhysicalNode]`. As this is an internal method, I think it's find to have an assert here to reduce the duplicate codes. 

For other methods in this class, I changed to `StreamPhysicalRel` parameter.

wuchong commented:
Thanks for the detailed review @godfreyhe and @libenchao . I have addressed all the comments. 
Please have a look again when you are free.

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/trait/ModifyKindSet.java:
```diff
@@ -0,0 +1,192 @@
+	 * [I,U] diff [U,D] = [I]
+	 * [I,U,D] diff [I,U,D] = []
+	 */
+	public ModifyKindSet diff(ModifyKindSet other) {
```
use `minus`? `diff` sounds like another meaning

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/trait/ModifyKindSet.java:
```diff
@@ -0,0 +1,192 @@
+
+	@Override
+	public String toString() {
+		if (kinds.isEmpty()) {
```
I think we should protect this class to let it be always non-empty.
Does an empty `ModifyKindSet` make any sense?

wuchong replied:
This is on purpose. Sink will produce emtpy `ModifyKindSet`, because it output nothing.

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/trait/UpdateKind.java:
```diff
@@ -0,0 +1,60 @@
+	/**
+	 * Return a new UpdateKind which merges this UpdateKind and the other UpdateKind.
+	 */
+	public UpdateKind merge(UpdateKind that) {
```
It's hard to imagine & understand why this class support `merge`, what should be the expected behavior?

wuchong replied:
Children of Union may have different UpdateKind. This is used to "union" children's UpdateKind. I was thinking "union" is not fit for this enum class, because it is not a bitset. 

I moved this logic into `FlinkChangelogModeInferenceProgram` as it is only used by StreamExecUnion.

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/trait/ModifyKindSetTrait.scala:
```diff
@@ -0,0 +1,77 @@
+      // e.g. [I,U] satisfy [I,U,D]
+      //      [I,U,D] not satisfy [I,D]
+      this.modifyKindSet.getContainedKinds.forall(other.modifyKindSet.contains)
+    case other: UpdateKindTrait =>
```
why `ModifyKindSetTrait` will be compared with `UpdateKindTrait`

wuchong replied:
You are right. This shouldn't be here. It is used to check whether the given UpdateKindTrait is valid. I move this logic into the validation in `FlinkChangelogModeInferenceProgram`.

KurtYoung commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/program/FlinkChangelogModeInferenceProgram.scala:
```diff
@@ -0,0 +1,626 @@
+            None
+        }
+
+      case _: StreamExecCalc | _: StreamExecPythonCalc | _: StreamExecCorrelate |
```
IIRC we've decided to not let `filter` passing the update kind. To solve the bad case like a single 'x < 10' condition which won't satisfy the `only_update_before` case.

wuchong replied:
Sorry, I forgot to implement this conclusion.

KurtYoung commented on flink-table/flink-table-planner-blink/src/test/resources/org/apache/flink/table/api/stream/ExplainTest.xml:
```diff
@@ -45,11 +45,11 @@ LogicalProject(EXPR$0=[$1])
-   +- Exchange(distribution=[hash[a]], updateAsRetraction=[true], accMode=[Acc]): rowcount = , cumulative cost = {rows, cpu, io, network, memory}
-      +- Calc(select=[a], updateAsRetraction=[true], accMode=[Acc]): rowcount = , cumulative cost = {rows, cpu, io, network, memory}
-         +- DataStreamScan(table=[[default_catalog, default_database, MyTable1]], fields=[a, b, c], updateAsRetraction=[true], accMode=[Acc]): rowcount = , cumulative cost = {rows, cpu, io, network, memory}
+Calc(select=[EXPR$0], changelogMode=[I,UA]): rowcount = , cumulative cost = {rows, cpu, io, network, memory}
```
why this operator is ignoring update before?

wuchong replied:
Because we are using `ONLY_UPDATE_AFTER` as the default required trait for queries without sink. This is set in [StreamCommonSubGraphBasedOptimizer.scala#L60](https://github.com/wuchong/flink/blob/9dcfd29d468b25f245e7b178ea4ae1a331ea6b39/flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/optimize/StreamCommonSubGraphBasedOptimizer.scala#L60) and passed by `StreamOptimizeContext#isUpdateBeforeRequired`.

wuchong commented:
Thanks for the reviewing. I have updated the branch to not satisfy ONLY_UPDATE_BEFORE for filters. 
And added tests in `SinkTest` and `TableSinkITCase`. Which should already fix FLINK-9528.

wuchong commented:
Merging...
